OpenSim/Simulation/Model/Smith2016Ligament.cpp

```diff
@@ -0,0 +1,377 @@
+	constructProperty_transition_strain(0.06);
+	constructProperty_normalized_damping_coefficient(0.003);
+	constructProperty_slack_length(0.0);
+	
+}
+
+void Smith2016Ligament::extendFinalizeFromProperties()
+{
+	Super::extendFinalizeFromProperties();
+
+	//Add channels to output list
+	auto& dyn_quan = updOutput("dynamic_quantities");
+	dyn_quan.addChannel("force_spring");
+	dyn_quan.addChannel("force_damping");
+	dyn_quan.addChannel("force_total");
+	dyn_quan.addChannel("length");
+	dyn_quan.addChannel("lengthening_speed");
+	dyn_quan.addChannel("strain");
+	dyn_quan.addChannel("strain_rate");	
+
+
+	//Check that properties are valid
+	OPENSIM_THROW_IF_FRMOBJ(get_slack_length() < 0.,
+        InvalidPropertyValue, getProperty_slack_length().getName(),
+        "Slack Length cannot be less than 0");
+
+	OPENSIM_THROW_IF_FRMOBJ(get_linear_stiffness() < 0.,
+        InvalidPropertyValue, getProperty_linear_stiffness().getName(),
+        "Linear Stiffness cannot be less than 0");
+
+	OPENSIM_THROW_IF_FRMOBJ(get_normalized_damping_coefficient() < 0.,
+        InvalidPropertyValue, getProperty_normalized_damping_coefficient().getName(),
+        "Normalized Damping Coefficient cannot be less than 0");
+
+	OPENSIM_THROW_IF_FRMOBJ(get_transition_strain() < 0.,
+        InvalidPropertyValue, getProperty_transition_strain().getName(),
+        "Transistion Strain cannot be less than 0");
+
+	//Set Ligament Color
+	GeometryPath& path = upd_GeometryPath();
+    path.setDefaultColor(SimTK::Vec3(0.99, 0.42, 0.01));
+
+}
+
+void Smith2016Ligament::extendAddToSystem(SimTK::MultibodySystem& system) const
+{
+	Super::extendAddToSystem(system);
+
+	addCacheVariable<double>("force_spring",0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("force_damping", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("force_total", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("length", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("lengthening_speed", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("strain", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("strain_rate", 0.0, SimTK::Stage::Dynamics);
+}
+
+
+void Smith2016Ligament::setSlackLengthFromReferenceStrain(double reference_strain, SimTK::State state) {
+	double reference_length = computeReferenceLength(state);
+	double slack_length = reference_length / (1.0 + reference_strain);
+};
+
+void Smith2016Ligament::setSlackLengthFromReferenceForce(double reference_force, SimTK::State state) {
+	double reference_strain = computeSpringStrain(reference_force);
+	setSlackLengthFromReferenceStrain(reference_strain, state);
+};
+
+double Smith2016Ligament::computeReferenceLength(SimTK::State state) const {
+
+	SimTK::Stage stage = state.getSystemStage();
+
+	
+	//Pose model to reference coordinates
+	for (Coordinate coord : getModel().getComponentList<Coordinate>()) {
+		coord.setValue(state, coord.getDefaultValue());
+	}
+	
+	getModel().realizePosition(state);
```

Comment:
The idea we discussed is that you would use the state the user provided, not try to edit the state to the default pose.